File: Javascript/Array/M-155. Min Stack.ts
class MinStack {
  valStack: number[];
  minValStack: number[];
  constructor() {
    this.valStack = [];
    this.minValStack = [];
  }

  push(val: number): void {
    /** valStack */
    this.valStack.push(val);

    /** minValStack */
    const n = this.minValStack.length;
    if (n === 0 || val < this.minValStack[n - 1]) {
      this.minValStack.push(val);
    } else {
      this.minValStack.push(this.minValStack[n - 1]);
    }
  }

  pop(): void {
    this.valStack.pop();
    this.minValStack.pop();
  }

  top(): number {
    const n = this.valStack.length;
    return this.valStack[n - 1];
  }

  getMin(): number {
    const n = this.minValStack.length;
    return this.minValStack[n - 1];
  }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: Array/M-155. Min Stack/M-155. Min Stack.py
class MinStack:

    def __init__(self):
        self.val_stack = []
        self.min_val_stack = []

    def push(self, val: int) -> None:
        # val_stack
        self.val_stack.append(val)

        # min_val_stack
        n = len(self.min_val_stack)
        if n == 0 or val < self.min_val_stack[n - 1]:
            self.min_val_stack.append(val)
        else:
            self.min_val_stack.append(self.min_val_stack[n-1])

    def pop(self) -> None:
        self.val_stack.pop()
        self.min_val_stack.pop()

    def top(self) -> int:
        return self.val_stack[len(self.val_stack) - 1]

    def getMin(self) -> int:
        return self.min_val_stack[len(self.min_val_stack) - 1]

        # Your MinStack object will be instantiated and called as such:
        # obj = MinStack()
        # obj.push(val)
        # obj.pop()
        # param_3 = obj.top()
        # param_4 = obj.getMin()
